crates/nu-command/src/core_commands/do_.rs

```diff
@@ -95,87 +107,89 @@ impl Command for Do {
             block,
             input,
             call.redirect_stdout,
-            ignore_errors || capture_errors,
+            capture_errors || ignore_shell_errors || ignore_program_errors,
         );
 
-        if ignore_errors {
-            match result {
-                Ok(x) => Ok(x),
-                Err(_) => Ok(PipelineData::new(call.head)),
-            }
-        } else if capture_errors {
-            // collect stdout and stderr and check exit code.
-            // if exit code is not 0, return back ShellError.
-            match result {
-                Ok(PipelineData::ExternalStream {
-                    stdout,
-                    stderr,
-                    exit_code,
-                    span,
-                    metadata,
-                }) => {
-                    // collect all output first.
-                    let mut stderr_ctrlc = None;
-                    let stderr_msg = match stderr {
-                        None => "".to_string(),
-                        Some(stderr_stream) => {
-                            stderr_ctrlc = stderr_stream.ctrlc.clone();
-                            stderr_stream.into_string().map(|s| s.item)?
-                        }
-                    };
-
-                    let mut stdout_ctrlc = None;
-                    let stdout_msg = match stdout {
-                        None => "".to_string(),
-                        Some(stdout_stream) => {
-                            stdout_ctrlc = stdout_stream.ctrlc.clone();
-                            stdout_stream.into_string().map(|s| s.item)?
-                        }
-                    };
-
-                    let mut exit_code_ctrlc = None;
-                    let exit_code: Vec<Value> = match exit_code {
-                        None => vec![],
-                        Some(exit_code_stream) => {
-                            exit_code_ctrlc = exit_code_stream.ctrlc.clone();
-                            exit_code_stream.into_iter().collect()
-                        }
-                    };
-                    if let Some(Value::Int { val: code, .. }) = exit_code.last() {
-                        // if exit_code is not 0, it indicates error occured, return back Err.
-                        if *code != 0 {
-                            return Err(ShellError::ExternalCommand(
-                                "External command runs to failed".to_string(),
-                                stderr_msg,
-                                span,
-                            ));
-                        }
+        match result {
+            Ok(PipelineData::ExternalStream {
+                stdout,
+                stderr,
+                exit_code,
+                span,
+                metadata,
+            }) if capture_errors => {
+                // collect all output first.
```

Comment:
If it appeals to you, I'd be in favor of breaking this block out into a separate function to make the top-level more readable. But totally up to you.

Reply:
I'd rather not pull out a method since we'll end up with a gazillion params, but I simplified the logic quite a bit.